contractor/processing_chain.cpp

```diff
@@ -100,12 +105,45 @@ int Prepare::Run()
     return 0;
 }
 
-std::size_t Prepare::LoadEdgeExpandedGraph(
-                std::string const & edge_based_graph_filename, 
-                DeallocatingVector<EdgeBasedEdge> & edge_based_edge_list)
+namespace std
+{
+
+template <> struct hash<std::pair<unsigned, unsigned>>
+{
+    std::size_t operator()(const std::pair<unsigned, unsigned> &k) const
+    {
+        return k.first ^ (k.second << 1);
+    }
+};
+}
+
+std::size_t Prepare::LoadEdgeExpandedGraph(std::string const &edge_based_graph_filename,
+                                           DeallocatingVector<EdgeBasedEdge> &edge_based_edge_list,
+                                           const std::string &edge_segment_lookup_filename,
+                                           const std::string &edge_penalty_filename,
+                                           const std::string &segment_speed_filename)
 {
     SimpleLogger().Write() << "Opening " << edge_based_graph_filename;
-    boost::filesystem::ifstream input_stream(edge_based_graph_filename, std::ios::in | std::ios::binary);
+    boost::filesystem::ifstream input_stream(edge_based_graph_filename,
+                                             std::ios::in | std::ios::binary);
+
+    bool update_edge_weights = segment_speed_filename != "";
+
+    boost::filesystem::ifstream edge_segment_input_stream;
+    boost::filesystem::ifstream edge_fixed_penalties_input_stream;
+
+    if (update_edge_weights)
+    {
+        edge_segment_input_stream.open(edge_segment_lookup_filename,
+                                       std::ios::in | std::ios::binary);
+        edge_fixed_penalties_input_stream.open(edge_penalty_filename,
+                                               std::ios::in | std::ios::binary);
+        if (!edge_segment_input_stream.is_open() || !edge_fixed_penalties_input_stream.is_open())
```

Comment:
instead of checking for `is_open`, I would just use the implicit conversion to bool that also checks for failues, like:

    if (not stream) { /* oops */ }